File: h/house-robber.cc
class Solution {
public:
  int rob(vector<int>& nums) {
    int n = nums.size();
    vector<int> d(n + 1, 0);

    for (int i = 1; i <= n; i++) {
      d[i] = max(d[i - 1], (i >= 2 ? d[i - 2] : 0) + nums[i - 1]);
    }

    return d[n];
  }
};

File: h/house-robber.rb
def rob_util(pos, nums, cache)
  return cache[pos] if cache[pos]

  if nums.size == pos
    result = 0
  elsif nums.size == pos + 1
    result = nums.last
  elsif nums.size == pos + 2
    result = nums.last(2).max
  else
    result = [
      rob_util(pos + 1, nums, cache),
      nums[pos] + rob_util(pos + 2, nums, cache)
    ].max
  end

  cache[pos] = result
end

# @param {Integer[]} nums
# @return {Integer}
def rob(nums)
  rob_util(0, nums, {})
end
